File: ios/Classes/EspIdfPlugin.swift
import Flutter
import UIKit
import ESPProvision
import CoreLocation

public class EspIdfPlugin: NSObject, FlutterPlugin {
    private var bleDeviceEventSink: FlutterEventSink?
    private var provisionManager: ESPProvisionManager?
    private var devices: [String: ESPDevice] = [:]
    private var bleDevices: [String: ESPDevice] = [:]
    private var pendingQRResult: FlutterResult?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "esp_idf", binaryMessenger: registrar.messenger())
        let instance = EspIdfPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)

        let bleDeviceEventChannel = FlutterEventChannel(
            name: "esp_idf/ble_devices",
            binaryMessenger: registrar.messenger()
        )
        bleDeviceEventChannel.setStreamHandler(BleDeviceStreamHandler(plugin: instance))

        let connectionEventChannel = FlutterEventChannel(
            name: "esp_idf/connection_status",
            binaryMessenger: registrar.messenger()
        )
        connectionEventChannel.setStreamHandler(ConnectionStreamHandler())

        let provisioningEventChannel = FlutterEventChannel(
            name: "esp_idf/provisioning_status",
            binaryMessenger: registrar.messenger()
        )
        provisioningEventChannel.setStreamHandler(ProvisioningStreamHandler())
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "searchBleDevices":
            searchBleDevices(call: call, result: result)
        case "stopBleSearch":
            stopBleSearch(result: result)
        case "createDevice":
            createDevice(call: call, result: result)
        case "connectDevice":
            connectDevice(call: call, result: result)
        case "disconnectDevice":
            disconnectDevice(call: call, result: result)
        case "scanWifiNetworks":
            scanWifiNetworks(call: call, result: result)
        case "provisionDevice":
            provisionDevice(call: call, result: result)
        case "isBluetoothEnabled":
            isBluetoothEnabled(result: result)
        case "requestPermissions":
            requestPermissions(result: result)
        case "checkPermissions":
            checkPermissions(result: result)
        case "scanQRCode":
            scanQRCode(call: call, result: result)
        case "connectWiFiDevice":
            connectWiFiDevice(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func searchBleDevices(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let prefix = args["prefix"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        provisionManager = ESPProvisionManager.shared

        provisionManager?.searchESPDevices(devicePrefix: prefix, transport: .ble, security: .secure) { foundDevices, error in
            if let error = error {
                self.bleDeviceEventSink?(FlutterError(
                    code: "SCAN_ERROR",
                    message: error.localizedDescription,
                    details: nil
                ))
                return
            }

            if let foundDevices = foundDevices {
                for device in foundDevices {
                    self.bleDevices[device.name] = device
                    let deviceMap: [String: Any] = [
                        "name": device.name,
                        "transport": "ble",
                        "security": "secure",
                        "deviceId": device.name
                    ]
                    self.bleDeviceEventSink?(deviceMap)
                }
            }
        }

        result(nil)
    }

    private func stopBleSearch(result: @escaping FlutterResult) {
        provisionManager?.stopESPDevicesSearch()
        result(nil)
    }

    private func createDevice(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let deviceName = args["deviceName"] as? String,
              let transportStr = args["transport"] as? String,
              let securityStr = args["security"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        let transport: ESPTransport = transportStr == "ble" ? .ble : .softap
        let security: ESPSecurity

        switch securityStr {
        case "unsecured":
            security = .unsecure
        case "secure":
            security = .secure
        case "secure2":
            security = .secure2
        default:
            security = .secure
        }

        let proofOfPossession = args["proofOfPossession"] as? String
        let username = args["username"] as? String
        let softApPassword = args["softApPassword"] as? String

        let device = ESPDevice(name: deviceName, security: security, transport: transport, proofOfPossession: proofOfPossession, username: username, softAPPassword: softApPassword)

        let deviceId = deviceName
        devices[deviceId] = device

        result(deviceId)
    }

    private func connectDevice(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let deviceId = args["deviceId"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        guard let device = devices[deviceId] ?? bleDevices[deviceId] else {
            result(FlutterError(code: "DEVICE_NOT_FOUND", message: "Device not found", details: nil))
            return
        }

        device.connect(delegate: self) { status in
            switch status {
            case .connected:
                result(nil)
            case let .failedToConnect(error):
                result(FlutterError(
                    code: "CONNECT_FAILED",
                    message: error.localizedDescription,
                    details: nil
                ))
            default:
                break
            }
        }
    }

    private func disconnectDevice(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let deviceId = args["deviceId"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        guard let device = devices[deviceId] ?? bleDevices[deviceId] else {
            result(FlutterError(code: "DEVICE_NOT_FOUND", message: "Device not found", details: nil))
            return
        }

        device.disconnect()
        result(nil)
    }

    private func scanWifiNetworks(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let deviceId = args["deviceId"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        guard let device = devices[deviceId] ?? bleDevices[deviceId] else {
            result(FlutterError(code: "DEVICE_NOT_FOUND", message: "Device not found", details: nil))
            return
        }

        device.scanWifiList { wifiList, error in
            if let error = error {
                result(FlutterError(
                    code: "SCAN_FAILED",
                    message: error.localizedDescription,
                    details: nil
                ))
                return
            }

            if let wifiList = wifiList {
                let networks = wifiList.map { wifi -> [String: Any] in
                    return [
                        "ssid": wifi.ssid,
                        "rssi": wifi.rssi,
                        "auth": wifi.auth.rawValue,
                        "channel": wifi.channel
                    ]
                }
                result(networks)
            } else {
                result([])
            }
        }
    }

    private func provisionDevice(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let deviceId = args["deviceId"] as? String,
              let ssid = args["ssid"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        let password = args["password"] as? String ?? ""

        guard let device = devices[deviceId] ?? bleDevices[deviceId] else {
            result(FlutterError(code: "DEVICE_NOT_FOUND", message: "Device not found", details: nil))
            return
        }

        device.provision(ssid: ssid, passPhrase: password) { status in
            switch status {
            case .success:
                result(nil)
            case let .failure(error):
                result(FlutterError(
                    code: "PROVISION_FAILED",
                    message: error.localizedDescription,
                    details: nil
                ))
            case .configApplied:
                // Config applied
                break
            }
        }
    }

    private func isBluetoothEnabled(result: @escaping FlutterResult) {
        // iOS doesn't provide a direct way to check Bluetooth state without permissions
        // Return true as a default
        result(true)
    }

    private func checkPermissions(result: @escaping FlutterResult) {
        // Check if location permission is granted (required for Wi-Fi SSID access)
        let status = CLLocationManager.authorizationStatus()
        let granted = status == .authorizedWhenInUse || status == .authorizedAlways
        result(granted)
    }

    private func requestPermissions(result: @escaping FlutterResult) {
        // Permissions need to be requested from the app level
        // Return true to indicate that the method was called
        // Actual permission requests should be handled by the app
        result(true)
    }

    private func scanQRCode(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let rootViewController = UIApplication.shared.windows.first?.rootViewController else {
            result(FlutterError(
                code: "NO_VIEW_CONTROLLER",
                message: "Unable to get root view controller",
                details: nil
            ))
            return
        }

        pendingQRResult = result

        let scannerVC = ESPQRScannerViewController()
        scannerVC.delegate = self

        // Apply custom parameters if provided
        if let args = call.arguments as? [String: Any] {
            if let title = args["title"] as? String {
                scannerVC.customTitle = title
            }
            if let description = args["description"] as? String {
                scannerVC.customDescription = description
            }
            if let cancelButtonText = args["cancelButtonText"] as? String {
                scannerVC.customCancelButtonText = cancelButtonText
            }
        }

        scannerVC.modalPresentationStyle = .fullScreen

        rootViewController.present(scannerVC, animated: true)
    }

    private func connectWiFiDevice(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let deviceId = args["deviceId"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Invalid arguments", details: nil))
            return
        }

        guard let device = devices[deviceId] ?? bleDevices[deviceId] else {
            result(FlutterError(code: "DEVICE_NOT_FOUND", message: "Device not found", details: nil))
            return
        }

        // For SoftAP transport, the connection is established when connected to the device's WiFi
        // This method just initiates the session
        device.connect(delegate: self) { status in
            switch status {
            case .connected:
                result(nil)
            case let .failedToConnect(error):
                result(FlutterError(
                    code: "CONNECT_FAILED",
                    message: error.localizedDescription,
                    details: nil
                ))
            default:
                break
            }
        }
    }

    func setBleDeviceEventSink(_ sink: FlutterEventSink?) {
        bleDeviceEventSink = sink
    }
}

// MARK: - ESPDeviceConnectionDelegate
extension EspIdfPlugin: ESPDeviceConnectionDelegate {
    public func getProofOfPossesion(forDevice: ESPDevice, completionHandler: @escaping (String) -> Void) {
        // In ESPProvision 3.0.3, proof of possession is passed in the initializer
        // We don't need to provide it again here, return empty string
        completionHandler("")
    }

    public func getUsername(forDevice: ESPDevice, completionHandler: @escaping (String?) -> Void) {
        // In ESPProvision 3.0.3, username is passed in the initializer
        // We don't need to provide it again here
        completionHandler(nil)
    }
}

// MARK: - ESPQRScannerDelegate
extension EspIdfPlugin: ESPQRScannerDelegate {
    func qrScannerDidScan(device: ESPDevice) {
        // Store the device
        let deviceId = device.name
        devices[deviceId] = device
        bleDevices[deviceId] = device

        // Determine transport and security
        let transport = device.transport == .ble ? "ble" : "softAp"
        let securityStr: String
        switch device.security {
        case .unsecure:
            securityStr = "unsecured"
        case .secure:
            securityStr = "secure"
        case .secure2:
            securityStr = "secure2"
        @unknown default:
            securityStr = "secure"
        }

        // Return device info to Flutter
        let deviceMap: [String: Any] = [
            "name": device.name,
            "transport": transport,
            "security": securityStr,
            "deviceId": deviceId
        ]

        pendingQRResult?(deviceMap)
        pendingQRResult = nil
    }

    func qrScannerDidCancel() {
        pendingQRResult?(FlutterError(
            code: "USER_CANCELLED",
            message: "QR code scanning was cancelled",
            details: nil
        ))
        pendingQRResult = nil
    }

    func qrScannerDidFail(error: Error) {
        pendingQRResult?(FlutterError(
            code: "SCAN_FAILED",
            message: error.localizedDescription,
            details: nil
        ))
        pendingQRResult = nil
    }
}

// MARK: - Stream Handlers
class BleDeviceStreamHandler: NSObject, FlutterStreamHandler {
    weak var plugin: EspIdfPlugin?

    init(plugin: EspIdfPlugin) {
        self.plugin = plugin
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        plugin?.setBleDeviceEventSink(events)
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        plugin?.setBleDeviceEventSink(nil)
        return nil
    }
}

class ConnectionStreamHandler: NSObject, FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
}

class ProvisioningStreamHandler: NSObject, FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
}

File: android/src/main/kotlin/com/nghinv/esp_idf/EspIdfPlugin.kt
package com.nghinv.esp_idf

import android.Manifest
import android.app.Activity
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothManager
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.espressif.provisioning.ESPConstants
import com.espressif.provisioning.ESPDevice
import com.espressif.provisioning.ESPProvisionManager
import com.espressif.provisioning.WiFiAccessPoint
import com.espressif.provisioning.listeners.BleScanListener
import com.espressif.provisioning.listeners.ProvisionListener
import com.espressif.provisioning.listeners.WiFiScanListener
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry
import java.util.UUID

/** EspIdfPlugin */
class EspIdfPlugin : FlutterPlugin, MethodCallHandler, ActivityAware,
    PluginRegistry.RequestPermissionsResultListener {

    private lateinit var channel: MethodChannel
    private lateinit var bleDeviceEventChannel: EventChannel
    private lateinit var connectionEventChannel: EventChannel
    private lateinit var provisioningEventChannel: EventChannel

    private var bleDeviceEventSink: EventChannel.EventSink? = null
    private var provisionManager: ESPProvisionManager? = null
    private val devices = mutableMapOf<String, ESPDevice>()
    private var activity: Activity? = null
    private var context: Context? = null
    private var pendingPermissionResult: Result? = null
    private var pendingQRResult: Result? = null

    companion object {
        private const val PERMISSION_REQUEST_CODE = 1001
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "esp_idf")
        channel.setMethodCallHandler(this)

        bleDeviceEventChannel = EventChannel(
            flutterPluginBinding.binaryMessenger,
            "esp_idf/ble_devices"
        )
        bleDeviceEventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                bleDeviceEventSink = events
            }

            override fun onCancel(arguments: Any?) {
                bleDeviceEventSink = null
            }
        })

        connectionEventChannel = EventChannel(
            flutterPluginBinding.binaryMessenger,
            "esp_idf/connection_status"
        )

        provisioningEventChannel = EventChannel(
            flutterPluginBinding.binaryMessenger,
            "esp_idf/provisioning_status"
        )

        provisionManager = ESPProvisionManager.getInstance(context)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "searchBleDevices" -> searchBleDevices(call, result)
            "stopBleSearch" -> stopBleSearch(result)
            "createDevice" -> createDevice(call, result)
            "connectDevice" -> connectDevice(call, result)
            "disconnectDevice" -> disconnectDevice(call, result)
            "scanWifiNetworks" -> scanWifiNetworks(call, result)
            "provisionDevice" -> provisionDevice(call, result)
            "isBluetoothEnabled" -> isBluetoothEnabled(result)
            "requestPermissions" -> requestPermissions(result)
            "checkPermissions" -> checkPermissions(result)
            "scanQRCode" -> scanQRCode(result)
            "connectWiFiDevice" -> connectWiFiDevice(call, result)
            else -> result.notImplemented()
        }
    }

    private fun searchBleDevices(call: MethodCall, result: Result) {
        val prefix = call.argument<String>("prefix") ?: ""

        provisionManager?.searchBleEspDevices(prefix, object : BleScanListener {
            override fun scanStartFailed() {
                bleDeviceEventSink?.error("SCAN_FAILED", "Failed to start BLE scan", null)
            }

            override fun onPeripheralFound(device: BluetoothDevice?, scanResult: android.bluetooth.le.ScanResult?) {
                device?.let {
                    val deviceName = it.name ?: ""
                    val deviceMap = mapOf(
                        "name" to deviceName,
                        "transport" to "ble",
                        "security" to "secure",
                        "deviceId" to it.address
                    )
                    bleDeviceEventSink?.success(deviceMap)
                }
            }

            override fun scanCompleted() {
                // Scan completed
            }

            override fun onFailure(e: Exception?) {
                bleDeviceEventSink?.error("SCAN_ERROR", e?.message ?: "Unknown error", null)
            }
        })

        result.success(null)
    }

    private fun stopBleSearch(result: Result) {
        // The Android library doesn't have an explicit stop method
        // Scan automatically stops after timeout
        result.success(null)
    }

    private fun createDevice(call: MethodCall, result: Result) {
        val deviceName = call.argument<String>("deviceName") ?: ""
        val transport = call.argument<String>("transport") ?: "ble"
        val security = call.argument<String>("security") ?: "secure"
        val proofOfPossession = call.argument<String>("proofOfPossession")
        val username = call.argument<String>("username")
        val softApPassword = call.argument<String>("softApPassword")
        val primaryServiceUuid = call.argument<String>("primaryServiceUuid")

        val espTransport = when (transport) {
            "ble" -> ESPConstants.TransportType.TRANSPORT_BLE
            "softAp" -> ESPConstants.TransportType.TRANSPORT_SOFTAP
            else -> ESPConstants.TransportType.TRANSPORT_BLE
        }

        val espSecurity = when (security) {
            "unsecured" -> ESPConstants.SecurityType.SECURITY_0
            "secure" -> ESPConstants.SecurityType.SECURITY_1
            "secure2" -> ESPConstants.SecurityType.SECURITY_2
            else -> ESPConstants.SecurityType.SECURITY_1
        }

        val device = provisionManager?.createESPDevice(espTransport, espSecurity)

        device?.let {
            it.deviceName = deviceName
            proofOfPossession?.let { pop -> it.proofOfPossession = pop }
            username?.let { user -> it.userName = user }

            if (espTransport == ESPConstants.TransportType.TRANSPORT_BLE && primaryServiceUuid != null) {
                it.primaryServiceUuid = primaryServiceUuid
            }

            val deviceId = deviceName // Use device name as ID
            devices[deviceId] = it
            result.success(deviceId)
        } ?: result.error("CREATE_FAILED", "Failed to create device", null)
    }

    private fun connectDevice(call: MethodCall, result: Result) {
        val deviceId = call.argument<String>("deviceId") ?: ""
        val device = devices[deviceId]

        if (device == null) {
            result.error("DEVICE_NOT_FOUND", "Device not found", null)
            return
        }

        device.connectToDevice()
        result.success(null)
    }

    private fun disconnectDevice(call: MethodCall, result: Result) {
        val deviceId = call.argument<String>("deviceId") ?: ""
        val device = devices[deviceId]

        if (device == null) {
            result.error("DEVICE_NOT_FOUND", "Device not found", null)
            return
        }

        device.disconnectDevice()
        result.success(null)
    }

    private fun scanWifiNetworks(call: MethodCall, result: Result) {
        val deviceId = call.argument<String>("deviceId") ?: ""
        val device = devices[deviceId]

        if (device == null) {
            result.error("DEVICE_NOT_FOUND", "Device not found", null)
            return
        }

        device.scanNetworks(object : WiFiScanListener {
            override fun onWifiListReceived(wifiList: ArrayList<WiFiAccessPoint>?) {
                val networks = wifiList?.map { wifi ->
                    mapOf(
                        "ssid" to wifi.wifiName,
                        "rssi" to wifi.rssi,
                        "auth" to wifi.security,
                        "channel" to 0
                    )
                } ?: emptyList()

                result.success(networks)
            }

            override fun onWiFiScanFailed(e: Exception?) {
                result.error("SCAN_FAILED", e?.message ?: "Wi-Fi scan failed", null)
            }
        })
    }

    private fun provisionDevice(call: MethodCall, result: Result) {
        val deviceId = call.argument<String>("deviceId") ?: ""
        val ssid = call.argument<String>("ssid") ?: ""
        val password = call.argument<String>("password") ?: ""
        val device = devices[deviceId]

        if (device == null) {
            result.error("DEVICE_NOT_FOUND", "Device not found", null)
            return
        }

        device.provision(ssid, password, object : ProvisionListener {
            override fun createSessionFailed(e: Exception?) {
                result.error("SESSION_FAILED", e?.message ?: "Session creation failed", null)
            }

            override fun wifiConfigSent() {
                // Config sent
            }

            override fun wifiConfigFailed(e: Exception?) {
                result.error("CONFIG_FAILED", e?.message ?: "Config failed", null)
            }

            override fun wifiConfigApplied() {
                // Config applied
            }

            override fun wifiConfigApplyFailed(e: Exception?) {
                result.error("APPLY_FAILED", e?.message ?: "Apply failed", null)
            }

            override fun provisioningFailedFromDevice(failureReason: ESPConstants.ProvisionFailureReason?) {
                result.error("PROVISION_FAILED", "Provisioning failed: $failureReason", null)
            }

            override fun deviceProvisioningSuccess() {
                result.success(null)
            }

            override fun onProvisioningFailed(e: Exception?) {
                result.error("PROVISION_ERROR", e?.message ?: "Provisioning error", null)
            }
        })
    }

    private fun scanQRCode(result: Result) {
        if (activity == null) {
            result.error("NO_ACTIVITY", "Activity not available", null)
            return
        }

        // Store the result for later use when QR code is scanned
        pendingQRResult = result

        // For QR code scanning, we'll use a simpler approach
        // The app should handle QR scanning UI and pass the data to createDevice
        result.error("NOT_IMPLEMENTED",
            "QR scanning should be implemented in the Flutter UI layer. Use a package like mobile_scanner or qr_code_scanner, then pass the data to createDevice().",
            null)
    }

    private fun connectWiFiDevice(call: MethodCall, result: Result) {
        val deviceId = call.argument<String>("deviceId") ?: ""
        val device = devices[deviceId]

        if (device == null) {
            result.error("DEVICE_NOT_FOUND", "Device not found", null)
            return
        }

        // For SoftAP, we need to use connectWiFiDevice() instead of connectToDevice()
        device.connectWiFiDevice()
        result.success(null)
    }

    private fun isBluetoothEnabled(result: Result) {
        val bluetoothManager = context?.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
        val bluetoothAdapter = bluetoothManager?.adapter
        result.success(bluetoothAdapter?.isEnabled == true)
    }

    private fun checkPermissions(result: Result) {
        val permissions = getRequiredPermissions()
        val allGranted = permissions.all { permission ->
            ContextCompat.checkSelfPermission(context!!, permission) == PackageManager.PERMISSION_GRANTED
        }
        result.success(allGranted)
    }

    private fun requestPermissions(result: Result) {
        if (activity == null) {
            result.error("NO_ACTIVITY", "Activity not available", null)
            return
        }

        val permissions = getRequiredPermissions()
        val notGranted = permissions.filter { permission ->
            ContextCompat.checkSelfPermission(context!!, permission) != PackageManager.PERMISSION_GRANTED
        }

        if (notGranted.isEmpty()) {
            result.success(true)
            return
        }

        pendingPermissionResult = result
        ActivityCompat.requestPermissions(
            activity!!,
            notGranted.toTypedArray(),
            PERMISSION_REQUEST_CODE
        )
    }

    private fun getRequiredPermissions(): List<String> {
        val permissions = mutableListOf<String>()

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            permissions.add(Manifest.permission.BLUETOOTH_SCAN)
            permissions.add(Manifest.permission.BLUETOOTH_CONNECT)
        } else {
            permissions.add(Manifest.permission.BLUETOOTH)
            permissions.add(Manifest.permission.BLUETOOTH_ADMIN)
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            permissions.add(Manifest.permission.ACCESS_FINE_LOCATION)
        } else {
            permissions.add(Manifest.permission.ACCESS_COARSE_LOCATION)
            permissions.add(Manifest.permission.ACCESS_FINE_LOCATION)
        }

        permissions.add(Manifest.permission.ACCESS_WIFI_STATE)
        permissions.add(Manifest.permission.CHANGE_WIFI_STATE)
        permissions.add(Manifest.permission.ACCESS_NETWORK_STATE)
        permissions.add(Manifest.permission.CHANGE_NETWORK_STATE)

        return permissions
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ): Boolean {
        if (requestCode == PERMISSION_REQUEST_CODE) {
            val allGranted = grantResults.all { it == PackageManager.PERMISSION_GRANTED }
            pendingPermissionResult?.success(allGranted)
            pendingPermissionResult = null
            return true
        }
        return false
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        bleDeviceEventChannel.setStreamHandler(null)
        connectionEventChannel.setStreamHandler(null)
        provisioningEventChannel.setStreamHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addRequestPermissionsResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addRequestPermissionsResultListener(this)
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}
